main.cpp

```diff
@@ -12,147 +12,367 @@
 //    See the License for the specific language governing permissions and
 //    limitations under the License.
 
-#include <iostream>
-
-#include <libflash/fileio.hpp>
-#include <libflash/optimized_writer.hpp>
-#include <common/common.hpp>
+#define _GNU_SOURCE /* needed for splice() */
 
+#include <errno.h>
+#include <fcntl.h>
 #include <getopt.h>
+#include <mtd/ubi-user.h>
+#include <stdbool.h>
+#include <stdlib.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <string.h>
+#include <sys/ioctl.h>
+#include <sys/sendfile.h>
+#include <sys/stat.h>
+#include <sys/sysmacros.h>
+#include <unistd.h>
+
+#include "config.h"
+
+#define UBIMajorDevNo 10
+#define BLOCK_SIZE (1024 * 1024L) /* 1 MiB */
+#define MIN(X, Y) ((X < Y) ? X : Y)
+
+static struct option long_options[] = {
+	{"help", no_argument, 0, 'h'},
+	{"write-everything", no_argument, 0, 'w'},
+	{"input-size", required_argument, 0, 's'},
+	{"fsync-interval", required_argument, 0, 'f'},
+	{"input", required_argument, 0, 'i'},
+	{"output", required_argument, 0, 'o'},
+	{0, 0, 0, 0}};
 
 void PrintHelp() {
-	std::cout
-		<< "Usage: mender-flash [-h|--help] [-s|--input-size <INPUT_SIZE>] -i|--input <INPUT_PATH> -o|--output <OUTPUT_PATH>"
-		<< std::endl;
+	fputs(
+		"Usage:\n"
+		"  mender-flash [-h|--help] [-w|--write-everything] [-s|--input-size <INPUT_SIZE>] [-f|--fsync-interval <FSYNC_INTERVAL>] -i|--input <INPUT_PATH> -o|--output <OUTPUT_PATH>\n",
+		stderr);
 }
 
-int main(int argc, char *argv[]) {
-	size_t volumeSize = 0;
-	std::string inputPath;
-	std::string outputPath;
-	const size_t blockSize = 1024 * 1024; // 1MiB block size
-
-	while (true) {
-		static struct option long_options[] = {
-			{"help", no_argument, 0, 'h'},
-			{"input-size", required_argument, 0, 's'},
-			{"input", required_argument, 0, 'i'},
-			{"output", required_argument, 0, 'o'},
-			{0, 0, 0, 0}};
-
-		int option_index = 0;
-		int c = getopt_long(argc, argv, "hs:i:o:", long_options, &option_index);
-		if (c == -1) {
-			break;
+typedef ssize_t (*io_fn_t)(int, void *, size_t);
+
+struct Stats {
+	size_t blocks_written;
+	size_t blocks_omitted;
+	uint64_t bytes_written;
+	uint64_t bytes_omitted;
+	uint64_t total_bytes;
+};
+
+ssize_t buf_io(io_fn_t io_fn, int fd, unsigned char *buf, size_t len) {
+	size_t rem = len;
+	ssize_t n_done;
+	do {
+		n_done = io_fn(fd, buf + (len - rem), rem);
+		if (n_done > 0) {
+			rem -= n_done;
+		} else if ((n_done == -1) && (errno == EINTR)) {
+			continue;
 		}
+	} while ((n_done > 0) && (rem > 0) && (len > 0));
+
+	if (n_done < 0) {
+		return n_done;
+	} else {
+		return (len - rem);
+	}
+}
+
+bool shovel_data(
+	int in_fd,
+	int out_fd,
+	size_t len,
+	bool write_optimized,
+	size_t fsync_interval,
+	struct Stats *stats,
+	int *error) {
+	unsigned char buffer[BLOCK_SIZE];
+	size_t n_unsynced = 0;
```

Comment:
This is off-topic, but a perfect demonstration of why curly braces should be on a separate line. It would make this significantly more readable.

Reply:
I'd say it's a perfect demonstration of why no strict, always-applied code formatter results in nice code. Look at how nice and easy to read it was before the commit that unleashed the formatter on it.

Reply:
Besides that the readability argument is quite subjective, an automated formatter is exactly meant to make such useless discussions obsolete and it did in so many repositories.
This conversation once again shows that some people are just not willing to work in a team and adhering to an agreed on convention. This is hugely disappointing.

Reply:
> Sure, then there are discussion about what the right settings for the formatter are. And every change causes harm to git history etc.
...
> Anyway, if you want to continue with this discussion, which I'd love to, let's please take it somewhere else.

Yes and there is a place for this: contact the body in charge of the formatting we apply to the project or stop complaining and start contributing according to the agreed on formatting rules.

> Also, if by "some people" you mean me, take a look at this PR and all my commits in the mender (client) repo.

This discussion we can indeed take elsewhere although based on the experiences I've made so far I don't think there is any hope in getting to a not-20-year-old way of working and rather stick to other projects to make progress.

Reply:
This is not a discussion about whether or not to use a code formatter. Using a code formatter has a large majority vote in the team. Same with curly-braces-on-the-same-line, this was voted on in the fall of 2022, and won over curly-on-new-line. End of story.

Now, for the specific problem that Ole H. pointed out above, I agree that this could be improved. Two options I can think of are:

* Indent multi-line argument lists two levels instead of one.
* Insert a newline between the curly opening line and the first line of code.

Whoever cares enough about this, feel free to check if clang-format has an option for either of these, or something else sensible, and suggest it in a separate PR. But be prepared to follow up with fixes in the Mender repos as well, including the original mender-flash (*).

(*) Regardless of whether we decide to merge this rewrite PR, we will not merge it before the release, since we are in effective feature-freeze now.